### ISSUE TITLE
Training fails upon empty spectrum

### ISSUE THREAD

**Issue by guhanrv** [Noble-Lab/casanovo#16]
Hello again,

For some reason, the `on_train_epoch_end` hook is called before training is complete. I am trying to train with a pretty large dataset (~1.2M spectra between the training and validation sets, with 1093774 spectra (17,091 batches of 64) for training and 123930 spectra (1,937 batches of 64) for validation). For some reason, casanovo is thinking that training is done after only 11,927 batches (which is significantly less than the 17,091 expected) and so calls `on_train_epoch_end`, which looks for a history to add the train loss to and fails:

```
...
/home/ubuntu/smsnet_val_data/1049.mgf: 46spectra [00:00, 2628.76spectra/s]
/home/ubuntu/smsnet_val_data/1276.mgf: 62spectra [00:00, 2859.35spectra/s]
/home/ubuntu/smsnet_val_data/1073.mgf: 1spectra [00:00, 4156.89spectra/s]
/home/ubuntu/smsnet_val_data/1224.mgf: 4spectra [00:00, 5257.67spectra/s]
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
Global seed set to 454
initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/1
INFO: Added key: store_based_barrier_key:1 to store for rank: 0
INFO: Rank 0: Completed store-based barrier for key:store_based_barrier_key:1 with 1 nodes.
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All distributed processes registered. Starting with 1 processes
----------------------------------------------------------------------------------------------------

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name    | Type             | Params
---------------------------------------------
0 | encoder | SpectrumEncoder  | 18.9 M
1 | decoder | PeptideDecoder   | 28.4 M
2 | softmax | Softmax          | 0     
3 | celoss  | CrossEntropyLoss | 0     
---------------------------------------------
47.3 M    Trainable params
0         Non-trainable params
47.3 M    Total params
189.387   Total estimated model params size (MB)
/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:631: UserWarning: Checkpoint directory /home/ubuntu/checkpoints exists and is not empty.
  rank_zero_warn(f"Checkpoint directory {dirpath} exists and is not empty.")
Epoch 0:   0%|                                                                                                                                        | 0/19028 [00:00<?, ?it/s]/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/utilities/data.py:59: UserWarning: Trying to infer the `batch_size` from an ambiguous collection. The batch size we found is 64. To avoid any miscalculations, use `self.log(..., batch_size=batch_size)`.
  warning_cache.warn(
Epoch 0:   0%|                                                                                                                  | 1/19028 [00:01<6:18:18,  1.19s/it, loss=0.868]INFO: Reducer buckets have been rebuilt in this iteration.
Epoch 0:  50%|█████████████████████████████████████████████████████████▏                                                        | 9549/19028 [30:34<30:20,  5.21it/s, loss=1.03]Epoch 0:  50%|█████████████████████████████████████████████████████████▎                                                        | 9575/19028 [30:39<30:16,  5.20it/s, loss=1.04]
Epoch 0:  50%|█████████████████████████████████████████████████████████▎                                                        | 9576/19028 [30:40<30:16,  5.20it/s, loss=1.04]

Epoch 0:  50%|█████████████████████████████████████████████████████████▍                                                        | 9577/19028 [30:40<30:16,  5.20it/s, loss=1.03]
Epoch 0:  50%|█████████████████████████████████████████████████████████▍                                                        | 9582/19028 [30:41<30:15,  5.20it/s, loss=1.01]
Epoch 0:  50%|█████████████████████████████████████████████████████████▍                                                        | 9583/19028 [30:41<30:15,  5.20it/s, loss=1.01]
Epoch 0:  50%|█████████████████████████████████████████████████████████▍                                                        | 9584/19028 [30:41<30:14,  5.20it/s, loss=0.99]
Epoch 0:  50%|████████████████████████████████████████████████████████▉                                                        | 9585/19028 [30:42<30:14,  5.20it/s, loss=0.986]
Epoch 0:  50%|████████████████████████████████████████████████████████▉                                                        | 9586/19028 [30:42<30:14,  5.20it/s, loss=0.975]
Epoch 0:  63%|██████████████████████████████████████████████████████████████████████▏                                         | 11929/19028 [39:47<23:40,  5.00it/s, loss=0.932]Traceback (most recent call last):
  File "/home/ubuntu/mambaforge/envs/denovo/bin/casanovo", line 8, in <module>
    sys.exit(main())
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/casanovo/casanovo.py", line 32, in main
    train(train_data_path, val_data_path, model_path, config_path)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/casanovo/denovo/train_test.py", line 134, in train
    trainer.fit(model, train_loader.train_dataloader(), val_loader.val_dataloader())
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 740, in fit
    self._call_and_handle_interrupt(
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 685, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 777, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1199, in _run
    self._dispatch()
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1279, in _dispatch
    self.training_type_plugin.start_training(self)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 202, in start_training
    self._results = trainer.run_stage()
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1289, in run_stage
    return self._run_train()
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1319, in _run_train
    self.fit_loop.run()
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/loops/fit_loop.py", line 234, in advance
    self.epoch_loop.run(data_fetcher)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 151, in run
    output = self.on_run_end()
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 298, in on_run_end
    self.trainer.call_hook("on_train_epoch_end")
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1501, in call_hook
    output = model_fx(*args, **kwargs)
  File "/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/casanovo/denovo/model.py", line 425, in on_train_epoch_end
    self._history[-1]["train"] = train_loss
IndexError: list index out of range
Epoch 0:  63%|██████▎   | 11929/19028 [39:51<23:43,  4.99it/s, loss=0.932] 
```

I tried wrapping `self._history[-1]["train"] = train_loss` in a `try` and `except` where if the `self._history` is not of length at least 1, this step is skipped. However, somehow this skips validation entirely and goes to the next training epoch without validation:

```
/home/ubuntu/mambaforge/envs/denovo/lib/python3.9/site-packages/pytorch_lightning/utilities/data.py:59: UserWarning: Trying to infer the `batch_size` from an ambiguous collection. The batch size we found is 64. To avoid any miscalculations, use `self.log(..., batch_size=batch_size)`.
  warning_cache.warn(
Global seed set to 454                                                                                                                                                          
Epoch 0:   0%|                                                                                                          | 1/19028 [00:01<6:21:44,  1.20s/it, loss=3.5, lr=0.000]INFO: Reducer buckets have been rebuilt in this iteration.
Epoch 0:  63%|███████████████████████████████████████████████████████████████▎                                     | 11929/19028 [39:49<23:41,  4.99it/s, loss=2.17, lr=5.96e-5]INFO: ---------------------------------------------------------------------------------------------------------
INFO:   Epoch |   Train Loss  |  Valid Loss | Valid AA precision | Valid AA recall | Valid Peptide recall 
INFO: ---------------------------------------------------------------------------------------------------------
INFO:       0 |      2.398110 |      3.474908 |      0.000851 |      0.007873 |      0.000000 
Epoch 1:   4%|████▍                                                                                                  | 829/19028 [02:41<59:12,  5.12it/s, loss=2.22, lr=6.37e-5]
```

I also added the learning rate to this printout just to make sure it hadn't stopped learning or something funky with learning rate had made it stop training. I am running with the same warmup and max iteration configuration that you are running with (100000, 600000). Is there anything that you have encountered that would suggest where the problem is coming from?

**Comment by melihyilmaz** [comment]
Hi Guhan,

I haven't encountered an early stopping problem during training before and it's actually weird that model calls `` 11929th at iteration despite correctly recognizing there would be a total of 19028 iterations. Did you try training the model with a smaller data set, maybe a subsample of the one you're using, and if so did you observe early stopping behavior?

**Comment by guhanrv** [comment]
Hi @melihyilmaz,

I found the issue after some digging in the source code for pytorch lightning. 

In the `pytorch_lightning/loops/base.py"` file, I noticed that the `run()` function **excepts** a `StopIteration` which effectively tells the training loop to end:

```
def run(self, *args: Any, **kwargs: Any) -> T:
        """The main entry point to the loop.

        Will frequently check the :attr:`done` condition and calls :attr:`advance`
        until :attr:`done` evaluates to ``True``.

        Override this if you wish to change the default behavior. The default implementation is:

        Example::

            def run(self, *args, **kwargs):
                if self.skip:
                    return self.on_skip()

                self.reset()
                self.on_run_start(*args, **kwargs)

                while not self.done:
                    self.advance(*args, **kwargs)

                output = self.on_run_end()
                return output

        Returns:
            The output of :attr:`on_run_end` (often outputs collected from each step of the loop)
        """
        if self.skip:
            return self.on_skip()

        self.reset()

        self.on_run_start(*args, **kwargs)

        while not self.done:
            try:
                self.on_advance_start(*args, **kwargs)
                self.advance(*args, **kwargs)
                self.on_advance_end()
                self.restarting = False
            except StopIteration:
                break

        output = self.on_run_end()
        return output
```

I removed the `try`/`except` and got a `StopIteration` on a particular spectrum in the `.mgf` that looked like this:

```
BEGIN IONS
TITLE=782.5711
PEPMASS=1088.5186
CHARGE=3+
SCANS=5711
RTINSECONDS=134.84
SEQ=QASTSEESDEMPVPDSESVFVIPGSALLWR
2731.2097 7026.7
END IONS
```

Since I had turned spectrum preprocessing on in the `config.py` file, and the default for `max_mz` in the `preprocess_peaks` function is 2500, the input into the `_get_filter_intensity_mask` function was an empty intensity array. This then threw an IndexError - `min_intensity *= intensity[intensity_idx[-1]]: IndexError: index -1 is out of bounds for axis 0 with size 0`.

TLDR, I think that it would be nice to include some error catching for empty spectra post-processing if possible, especially since the pytorch-lightning module has such bad error-messaging! I will remove this spectrum from the file for now.

**Comment by melihyilmaz** [comment]
I see, thanks for raising this issue and letting us know. I'll level the issue open to add error handling in the future.

**Comment by bittremieux** [comment]
Fixed in #55.
